proof/refine/ARM/Schedule_R.thy

```diff
@@ -4210,6 +4209,1003 @@ lemma refillUnblockCheck_corres':
+  apply (prop_tac "sc_at scPtr s")
+   apply (fastforce dest: valid_objs_valid_sched_context_size
+                    simp: vs_all_heap_simps obj_at_kh_kheap_simps is_sc_obj_def)
+  apply (frule state_relation_pspace_relation)
+  apply (frule sc_at_cross; simp?)
+  apply (frule state_relation_sc_relation; simp?)
+  apply (subst head_insufficient_simp; simp?)
+  apply (subst refillHdInsufficient_simp; simp)
+  apply (frule refill_hd_relation)
+   apply (fastforce simp: vs_all_heap_simps valid_refills'_def opt_map_def obj_at_simps)
+  apply (clarsimp simp: obj_at_def sc_at_ppred_def obj_at'_def projectKOs minBudget_def refill_map_def
+                        MIN_BUDGET_def kernelWCETTicks_def opt_map_def vs_all_heap_simps)
+  done
+
+lemma refill_pop_head_no_fail:
+  "no_fail (\<lambda>s. (\<exists>sc n. kheap s sc_ptr = Some (Structures_A.SchedContext sc n))
+                \<and> pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) sc_ptr)
+           (refill_pop_head sc_ptr)"
+  apply (wpsimp simp: refill_pop_head_def get_refills_def get_sched_context_def
+                  wp: get_object_wp update_sched_context_no_fail)
+  apply (clarsimp simp: obj_at_def a_type_def vs_all_heap_simps is_sc_obj_def)
+  done
+
+lemma refill_pop_head_sched_context_at[wp]:
+  "refill_pop_head sc_ptr' \<lbrace>\<lambda>s. \<exists>sc n. kheap s sc_ptr = Some (Structures_A.SchedContext sc n)\<rbrace>"
+  apply (clarsimp simp: refill_pop_head_def)
+  apply (wpsimp wp: update_sched_context_wp)
+  done
+
+lemma non_overlapping_merge_refills_no_fail:
+  "no_fail (\<lambda>s. (\<exists>sc n. kheap s sc_ptr = Some (Structures_A.SchedContext sc n))
+                \<and> pred_map (\<lambda>cfg. scrc_refills cfg \<noteq> []) (sc_refill_cfgs_of s) sc_ptr)
+           (non_overlapping_merge_refills sc_ptr)"
+  apply (wpsimp wp: refill_pop_head_no_fail
+              simp: non_overlapping_merge_refills_def update_refill_hd_def)
+  done
+
+lemma non_overlapping_merge_refills_is_active_sc[wp]:
+  "non_overlapping_merge_refills sc_ptr \<lbrace>is_active_sc sc_ptr'\<rbrace>"
+  apply (clarsimp simp: non_overlapping_merge_refills_def update_refill_hd_def)
+  apply (rule hoare_seq_ext_skip, solves wpsimp)
+  apply (wpsimp wp: update_sched_context_wp)
+  apply (clarsimp simp: vs_all_heap_simps obj_at_def)
+  done
+
+crunches non_overlapping_merge_refills
+  for valid_objs[wp]: valid_objs
+
+lemma nonOverLappingMergeRefills_valid_refills'[wp]:
+  "nonOverlappingMergeRefills scPtr \<lbrace>valid_refills' scPtr\<rbrace>"
+  apply (wpsimp simp: nonOverlappingMergeRefills_def)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+definition head_insufficient_loop_measure where
+  "head_insufficient_loop_measure sc_ptr
+     \<equiv> measure (\<lambda>(_, s). case kheap s sc_ptr of Some (Structures_A.SchedContext sc _)
+                                             \<Rightarrow> (length (sc_refills sc)))"
```

Comment:
Yours is on the Refine level, mine is on the abstract level. If `corres_whileLoop_abs` were used for the proof of `refillUnblockCheck_corres`, then this measure could be used there, too, but there's probably very little benefit in doing that

Reply:
The point is that we have managed in duplicating essentially the same proofs for both abstract and concrete. They contain the same information for the purpose. We don’t want to have such duplications in maintaining because any change is needed in the future we would have to update both if we keep the both.

Reply:
When I wrote the `refillUnblockCheck_corres` proof, `corres_whileLoop_abs` didn’t exist, just to mention.

Reply:
Would you like me to rewrite the proof of `handleOverrunLoop_corres` with `corres_whileLoop`?

Reply:
Oh, `headInsufficientLoop_corres`, it would be

Reply:
I am not sure. What I think I know is that this measure probably deserves a more generic name, maybe we can name it to mean it’s about the length of refills.

And, since we have a crossing lemma for refill length, it probably makes sense to use the argument on the abstract side. So, we can do the consolidation later, not as a part of this PR. The next question is whether we should remove the `whileLoop` corres rule for termination on the concrete side so that we kind of force ourselves to do the termination on the abstract side. Maybe we don’t need to decide now.

Reply:
I am happy to rename the measure. Would you like me to do this as part of this PR? It was originally written for `headInsufficientLoop`, and it appears here only because I moved it.

I think `corres_whileLoop` is more general than `corres_whileleLoop_abs`: it seems possble for a concrete function to terminate, but to be in correspondence with an abstract function which does not terminate (roughly speaking). I cannot imagine that we'd come across such a situation in the l4v proofs, but it's worth mentioning.

Reply:
Actually, it is enough if we know what terminates on the concrete side terminates on the abstract side as well. So, that suggests we should base ourselves on the concrete side, I suppose? I think we can leave this PR as is on this issue and deal with it separately.

Reply:
Well, `corres_whileLoop_abs` is certainly proving more than what strictly needs to be proved, but this is done for convenience: Mitch had an attempt at `headInsufficientLoop_corres` with the concrete termination, and from memory, it was not all that pleasant. He had to introduce analogues of `refills_unat_sum` etc., which prompted me to think about abstract termination and no_fail. There's not much difference between abstract and concrete in the other loops we have, but my preference would be to have more of these general rules available, and to then select the best one for the proof at hand

Reply:
So, it is the concrete version of `head_insufficient_length_at_least_two` that is hard to show, I suppose? It is quite obvious that `nonOverlappingMergeRefills` is refill length decreasing. And I see why he wanted to have crossing there. And *if* that had worked, one of my refill-length-decreasing proofs could have been used. We could be generally proving the “refill-length-decreasing” property for functions. Maybe. It is a bit surprising that there was no other way to cross that information.

This is not specifically about the case here, but the potential problem is that those more specific rules are actually less general, and more often than not things go in the opposite way - people pick their rule to guide the proof, and if the proof you happened to pick is not general enough, it might lead to a hard point or a dead-end. Also, the search process becomes harder and more costly with many similar but specialised rules. I guess this is an argument for having a smaller set of more general rules.